File: src/models/hand.ts
import { Card, Rank, Suit, isChiChi, isRedCard } from './card';

export interface Hand {
  readonly cards: Card[];
}

export function createHand(): Hand {
  return { cards: [] };
}

export function addCard(hand: Hand, card: Card): Hand {
  return { cards: [...hand.cards, card] };
}

export function removeCard(hand: Hand, cardId: string): Hand {
  return {
    cards: hand.cards.filter(c => c.id !== cardId),
  };
}

export function hasCard(hand: Hand, cardId: string): boolean {
  return hand.cards.some(c => c.id === cardId);
}

export function findCard(hand: Hand, cardId: string): Card | undefined {
  return hand.cards.find(c => c.id === cardId);
}

export function sortHand(hand: Hand): Hand {
  const suitOrder: Record<Suit, number> = {
    [Suit.Van]: 0,
    [Suit.Van2]: 1,
    [Suit.Sach]: 2,
  };

  const sorted = [...hand.cards].sort((a, b) => {
    if (a.rank !== b.rank) return a.rank - b.rank;
    return suitOrder[a.suit] - suitOrder[b.suit];
  });

  return { cards: sorted };
}

export function handSize(hand: Hand): number {
  return hand.cards.length;
}

// Count how many cards of a specific rank the player holds
export function countByRank(hand: Hand, rank: Rank): number {
  return hand.cards.filter(c => c.rank === rank).length;
}

// Count how many cards of a specific suit the player holds
export function countBySuit(hand: Hand, suit: Suit): number {
  return hand.cards.filter(c => c.suit === suit).length;
}

// Get all cards of a specific rank
export function cardsOfRank(hand: Hand, rank: Rank): Card[] {
  return hand.cards.filter(c => c.rank === rank);
}

// Get all cards of a specific suit
export function cardsOfSuit(hand: Hand, suit: Suit): Card[] {
  return hand.cards.filter(c => c.suit === suit);
}

// Count red score cards (for "8 Đỏ" cước)
export function countRedScoreCards(hand: Hand): number {
  return hand.cards.filter(c => isRedCard(c)).length;
}

File: src/ChanDanGian.Core/Models/Hand.cs
namespace ChanDanGian.Core.Models;

public record Hand(List<Card> Cards)
{
    public int Size => Cards.Count;

    public Hand AddCard(Card card) =>
        this with { Cards = new List<Card>(Cards) { card } };

    public Hand RemoveCard(string cardId) =>
        this with { Cards = Cards.Where(c => c.Id != cardId).ToList() };

    public Hand Sort()
    {
        var sorted = Cards
            .OrderBy(c => (int)c.Rank)
            .ThenBy(c => (int)c.Suit)
            .ToList();
        return this with { Cards = sorted };
    }

    public int CountByRank(Rank rank) =>
        Cards.Count(c => c.Rank == rank);

    public int CountBySuit(Suit suit) =>
        Cards.Count(c => c.Suit == suit);

    public List<Card> CardsOfRank(Rank rank) =>
        Cards.Where(c => c.Rank == rank).ToList();

    public int RedCardCount =>
        Cards.Count(c => c.IsRedCard);

    public int RedScoreCardCount =>
        Cards.Count(c => c.IsRedScoreCard);
}
